src/ota_image_libs/v1/resource_table/utils.py

```diff
@@ -121,35 +133,80 @@ def _thread_local_dctx(self) -> zstandard.ZstdDecompressor:
-        # NOTE: prevent the same bundle being prepared again and again
-        bundle_rsid = entry.filter_applied.list_resource_id()
+        filter_cfg = entry.filter_applied
+        assert isinstance(filter_cfg, BundleFilter)
+
+        # NOTE: only do one query for each bundle
+        bundle_rsid = filter_cfg.list_resource_id()
         with self._bundle_process_lock:
-            if bundle_rsid not in self._bundle:
-                bundle_entry = self._orm_pool.orm_select_entry(resource_id=bundle_rsid)
-                logger.debug(f"Requesting bundle({bundle_rsid=}): {bundle_entry}")
-                _bundle_save_tmp = self._download_dir / tmp_fname(str(bundle_rsid))
-                bundle_save_dst = self._download_dir / bundle_entry.digest.hex()
-
-                yield from self._prepare_resource(bundle_entry, _bundle_save_tmp)
-                os.replace(_bundle_save_tmp, bundle_save_dst)
-                self._bundle[bundle_rsid] = bundle_save_dst
-            bundle_fpath = self._bundle[bundle_rsid]
-
-        # NOTE: keep the bundle for later use, but if recreate from bundle failed, we delete the bundle
-        #       to let otaclient do the re-downloading.
+            if not (_bundle_info := self._bundle_per_locks.get(bundle_rsid)):
+                _bundle_entry = self._orm_pool.orm_select_entry(resource_id=bundle_rsid)
+                self._bundle_per_locks[bundle_rsid] = (
+                    _bundle_entry,
+                    _bundle_prepare_lock := PerBundleLock(),
+                    _bundle_ready_event := BundleReadyEvent(),
+                )
+            else:
+                _bundle_entry, _bundle_prepare_lock, _bundle_ready_event = _bundle_info
+
+        _bundle_f = self._download_dir / _bundle_entry.digest.hex()
+        # Only trigger upper to download resources when _bundle_ready_event is not set.
+        # To prevent other waiting threads just deadly waiting if the thread that does
+        #   the downloading failed and raised exception, other waiting threads will also
+        #   keep trying to get the _bundle_prepare_lock and prepare bundle.
+        for _ in range(MAX_ITERS_FOR_WAITING_BUNDLE):
+            if _bundle_ready_event.is_set():
+                break
+
+            if _bundle_prepare_lock.acquire(blocking=False):
+                try:
+                    yield from self._prepare_one_bundle(_bundle_f, _bundle_entry)
+                    _bundle_ready_event.set()
+                    break
+                except Exception as e:
+                    raise BundledRecreateFailed(
+                        f"failed to prepare bundle {_bundle_entry}: {e!r}"
+                    ) from e
+                finally:
+                    _bundle_prepare_lock.release()
+            time.sleep(WAITING_BUNDLE_INTERVAL)
+        else:
+            raise BundledRecreateFailed(
+                f"timeout waiting for bundle {_bundle_entry=} ready, will retry"
+            )
```

Comment:
Refinement to prevent meaninglessly waiting if download thread raises exceptions.